# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

In [23]:
# 加载包
import pandas as pd

# 加载数据
original_data = pd.read_csv('E:/file/data_analysis/python/小白玩转python/随课数据/数据评估和数据清洗篇/e_commerce.csv')
original_data.head(50)

InvoiceNo StockCode                          Description  ...  UnitPrice CustomerID         Country
0     536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER  ...       2.55    17850.0  United Kingdom
1     536365     71053                  WHITE METAL LANTERN  ...       3.39    17850.0  United Kingdom
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER  ...       2.75    17850.0  United Kingdom
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE  ...       3.39    17850.0  United Kingdom
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART.  ...       3.39    17850.0  United Kingdom
5     536365     22752         SET 7 BABUSHKA NESTING BOXES  ...       7.65    17850.0  United Kingdom
6     536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER  ...       4.25    17850.0  United Kingdom
7     536366     22633               HAND WARMER UNION JACK  ...       1.85    17850.0  United Kingdom
8     536366     22632            HAND WARMER RED POLKA DOT  ...       1.85    17850.0  United Kingdom
9     536367     84879        ASSORTED COLOUR BIRD ORNAMENT  ...       1.69    13047.0  United Kingdom
10    536367     22745           POPPY'S PLAYHOUSE BEDROOM   ...       2.10    13047.0  United Kingdom
11    536367     22748            POPPY'S PLAYHOUSE KITCHEN  ...       2.10    13047.0  United Kingdom
12    536367     22749    FELTCRAFT PRINCESS CHARLOTTE DOLL  ...       3.75    13047.0  United Kingdom
13    536367     22310              IVORY KNITTED MUG COSY   ...       1.65    13047.0  United Kingdom
14    536367     84969   BOX OF 6 ASSORTED COLOUR TEASPOONS  ...       4.25    13047.0  United Kingdom
15    536367     22623        BOX OF VINTAGE JIGSAW BLOCKS   ...       4.95    13047.0  United Kingdom
16    536367     22622       BOX OF VINTAGE ALPHABET BLOCKS  ...       9.95    13047.0  United Kingdom
17    536367     21754             HOME BUILDING BLOCK WORD  ...       5.95    13047.0  United Kingdom
18    536367     21755             LOVE BUILDING BLOCK WORD  ...       5.95    13047.0  United Kingdom
19    536367     21777          RECIPE BOX WITH METAL HEART  ...       7.95    13047.0  United Kingdom
20    536367     48187                  DOORMAT NEW ENGLAND  ...       7.95    13047.0  United Kingdom
21    536368     22960             JAM MAKING SET WITH JARS  ...       4.25    13047.0  United Kingdom
22    536368     22913          RED COAT RACK PARIS FASHION  ...       4.95    13047.0  United Kingdom
23    536368     22912       YELLOW COAT RACK PARIS FASHION  ...       4.95    13047.0  United Kingdom
24    536368     22914         BLUE COAT RACK PARIS FASHION  ...       4.95    13047.0  United Kingdom
25    536369     21756             BATH BUILDING BLOCK WORD  ...       5.95    13047.0  United Kingdom
26    536370     22728            ALARM CLOCK BAKELIKE PINK  ...       3.75    12583.0          France
27    536370     22727            ALARM CLOCK BAKELIKE RED   ...       3.75    12583.0          France
28    536370     22726           ALARM CLOCK BAKELIKE GREEN  ...       3.75    12583.0          France
29    536370     21724      PANDA AND BUNNIES STICKER SHEET  ...       0.85    12583.0          France
30    536370     21883                     STARS GIFT TAPE   ...       0.65    12583.0          France
31    536370     10002          INFLATABLE POLITICAL GLOBE   ...       0.85    12583.0          France
32    536370     21791   VINTAGE HEADS AND TAILS CARD GAME   ...       1.25    12583.0          France
33    536370     21035      SET/2 RED RETROSPOT TEA TOWELS   ...       2.95    12583.0          France
34    536370     22326  ROUND SNACK BOXES SET OF4 WOODLAND   ...       2.95    12583.0          France
35    536370     22629                  SPACEBOY LUNCH BOX   ...       1.95    12583.0          France
36    536370     22659              LUNCH BOX I LOVE LONDON  ...       1.95    12583.0          France
37    536370     22631             CIRCUS PARADE LUNCH BOX   ...       1.95    12583.0   

## 评估数据

### 评估整洁度

In [26]:

original_data.head(10) # 前10
original_data.sample(10) # 随机抽取10个数

InvoiceNo StockCode                          Description  ...  UnitPrice CustomerID         Country
222863    556459    84596B   SMALL DOLLY MIX DESIGN ORANGE BOWL  ...       0.83        NaN  United Kingdom
6404      536943     22295            HEART FILIGREE DOVE LARGE  ...       1.65    17884.0  United Kingdom
428157    573414     23424              GINGHAM RECIPE BOOK BOX  ...       4.95    16877.0  United Kingdom
192142    553390     21564       PINK  HEART SHAPE LOVE BUCKET   ...       2.95    17841.0  United Kingdom
381157    569846     23100        SILVER BELLS TABLE DECORATION  ...       1.25    13422.0  United Kingdom
101729    544930     22457      NATURAL SLATE HEART CHALKBOARD   ...       2.95    17472.0  United Kingdom
292502    562553     23309  SET OF 60 I LOVE LONDON CAKE CASES   ...       1.25        NaN  United Kingdom
338368    566539     23389               SPACEBOY MINI BACKPACK  ...       4.15    17228.0  United Kingdom
275575    560996     22352    LUNCH BOX WITH CUTLERY RETROSPOT   ...       2.55    17625.0  United Kingdom
207550    554993     22195         LARGE HEART MEASURING SPOONS  ...       1.65    17051.0  United Kingdom

[10 rows x 8 columns]

从抽样的结果来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看，每行代表一次商品交易，每列是交易的相关变量，因此不存在结构性问题

### 评估整洁干净度

In [28]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541909条观察值，其中`Description`和`CustomerID`变量有缺失值，InvoiceDate的数据类型为应为日期，CustomerID列的数据变量类型应为字符串，这两列需进行数据转换

### 评估缺失数据

In [29]:
# 找出Description列所有缺失的行
original_data[original_data['Description'].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  CustomerID         Country
622       536414     22139         NaN        56  12/1/2010 11:52        0.0         NaN  United Kingdom
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0         NaN  United Kingdom
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0         NaN  United Kingdom
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0         NaN  United Kingdom
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0         NaN  United Kingdom
...          ...       ...         ...       ...              ...        ...         ...             ...
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0         NaN  United Kingdom
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0         NaN  United Kingdom
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0         NaN  United Kingdom
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0         NaN  United Kingdom
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0         NaN  United Kingdom

[1454 rows x 8 columns]

从输出结果来看，这些缺失Description的行，对应的UnitPrice行全为0，为验证这些猜想，增加筛选条件，看是否存在Description行缺失时UnitPrice却不为零的情况

In [31]:
original_data[(original_data['Description'].isnull()) & (original_data['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

从上面的输出结果来看，说明这些缺失Description的行，对应的UnitPrice行全为0。   
Description表示商品名称，UnitPrice表示商品价格，都是后续进行商品交易分析的重要变量，如果他们同时缺失，我们认为数据无法提供有效含义，因此这些数据后续可以被删除

`CustomerID`也存在缺失值，因此先根据条件提取出观察值

In [39]:
original_data[original_data['CustomerID'].isnull()]

InvoiceNo StockCode                      Description  ...  UnitPrice CustomerID         Country
622       536414     22139                              NaN  ...       0.00        NaN  United Kingdom
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE  ...       2.51        NaN  United Kingdom
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE  ...       2.51        NaN  United Kingdom
1445      536544     21786               POLKADOT RAIN HAT   ...       0.85        NaN  United Kingdom
1446      536544     21787            RAIN PONCHO RETROSPOT  ...       1.66        NaN  United Kingdom
...          ...       ...                              ...  ...        ...        ...             ...
541536    581498    85099B          JUMBO BAG RED RETROSPOT  ...       4.13        NaN  United Kingdom
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE  ...       4.13        NaN  United Kingdom
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN  ...       4.96        NaN  United Kingdom
541539    581498     85174                S/4 CACTI CANDLES  ...      10.79        NaN  United Kingdom
541540    581498       DOT                   DOTCOM POSTAGE  ...    1714.17        NaN  United Kingdom

[135080 rows x 8 columns]

`CustomerID`表示客户编号，不是分析畅销产品的必要变量，而且从输出结果来看，虽然`CustomerID`为空，其他的数据仍有效，因此保留此变量为空的值

### 评估重复数据

根据变量的含义来看，虽然`InvoiceNo`、`StockCode`、`CustomerID`都是唯一标识符，但一次交易可能存在多个商品，所以`InvoiceNo`可以重复；不同交易可能包含同件商品，因此`StockCode`可以重复；顾客可以多次交易购买商品，因此`CustomerID`也可以重复。

### 评估不一致数据

不一致数据可能存在于`Country`变量中，我们要查看是否存在多个不同的值指代同一国家的情况

In [43]:
original_data['Country'].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


从`Country`变量值来看，'USA'、'United States '均表示美国，'United Kingdom'、'UK'、'U.K.'均表示英国，因此应该在数据清洗阶段对这些值进行统一，只保留一个代称。

### 评估无效或错误数据

可以通过`describe()`方法，对数值统计信息进行快速了解

In [44]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，`Quantity`和 `UnitPrice`存在负数，会对后续的分析产生影响，因此先筛选出`Quantity`数值为负数的观察值，进一步评估其含义

In [45]:
original_data[original_data['Quantity']<0]

InvoiceNo StockCode                       Description  ...  UnitPrice CustomerID         Country
141      C536379         D                          Discount  ...      27.50    14527.0  United Kingdom
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS  ...       4.65    15311.0  United Kingdom
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE   ...       1.65    17548.0  United Kingdom
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES   ...       0.29    17548.0  United Kingdom
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES   ...       0.29    17548.0  United Kingdom
...          ...       ...                               ...  ...        ...        ...             ...
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL  ...       0.83    14397.0  United Kingdom
541541   C581499         M                            Manual  ...     224.69    15498.0  United Kingdom
541715   C581568     21258        VICTORIAN SEWING BOX LARGE  ...      10.95    15311.0  United Kingdom
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER  ...       1.25    17315.0  United Kingdom
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT  ...       1.25    17315.0  United Kingdom

[10624 rows x 8 columns]

从删选条件来看，`Quantity`变量为负的观察值，`InvoiceNo`似乎均以C开头，表示订单被取消，后续进行验证此猜想

In [47]:
original_data[(original_data['Quantity']<0) & (original_data['InvoiceNo'].str[0] != 'C')]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  CustomerID         Country
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0         NaN  United Kingdom
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0         NaN  United Kingdom
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0         NaN  United Kingdom
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0         NaN  United Kingdom
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0         NaN  United Kingdom
...          ...       ...         ...       ...              ...        ...         ...             ...
535333    581210     23395       check       -26  12/7/2011 18:36        0.0         NaN  United Kingdom
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0         NaN  United Kingdom
535336    581213     22576       check       -30  12/7/2011 18:38        0.0         NaN  United Kingdom
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0         NaN  United Kingdom
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0         NaN  United Kingdom

[1336 rows x 8 columns]

上述猜想被证明错误，但从结果来看，这些删选出来观察值`UnitPrice`均为0，因此增加`UnitPrice`的条件进行验证

In [48]:
original_data[(original_data['Quantity']<0) & (original_data['InvoiceNo'].str[0] != 'C') & (original_data['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，当`Quantity`变量为负数的时候，观察值满足以下条件之一：
- `InvoiceNo`以C开头，表示订单被取消
- `UnitPrice`为0，说明单价为0英镑

这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，将`Quantity`为负数的观察值删除。

接下来筛选出`UnitPrice`为负数的行

In [50]:
original_data[original_data['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  UnitPrice  CustomerID         Country
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51  -11062.06         NaN  United Kingdom
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52  -11062.06         NaN  United Kingdom

从输出结果来看，`UnitPrice`为负数的观察值都是坏账调整，不属于实际的商品交易数据，因此也在数据清理的步骤中将其删除

## 清洗数据

根据前面评估的部分可得到以下结论，我们需进行的清理步骤包括
- 把`InvoiceDate`变量的数据类型转换为为日期时间
- 把`CustomerID`变量的数据类型转换为字符串
- 把`Description`变量缺失的观察值删除
- 把`Country`变量值`"USA"`替换为`"United States"`
- 把`Country`变量值`"UK"`、`"U.K."`替换为`"United Kingdom"`
- 把`Quantity`变量值为负数的观察值删除
- 把`UnitPrice`变量值为负数的观察值删除

为区分清理前后的数据，将清理后的数据命名为`cleaned_data`

In [53]:
# 拷贝原始变量

cleaned_date  = original_data.copy()
cleaned_date.head()

InvoiceNo StockCode                          Description  ...  UnitPrice CustomerID         Country
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER  ...       2.55    17850.0  United Kingdom
1    536365     71053                  WHITE METAL LANTERN  ...       3.39    17850.0  United Kingdom
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER  ...       2.75    17850.0  United Kingdom
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE  ...       3.39    17850.0  United Kingdom
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.  ...       3.39    17850.0  United Kingdom

[5 rows x 8 columns]

把`InvoiceDate`变量的数据类型转换为为日期时间

In [58]:
# 利用pandas中的to_datetime()函数进行时间格式调整
cleaned_date['InvoiceDate']  = pd.to_datetime(cleaned_date['InvoiceDate'])
cleaned_date['InvoiceDate'] 

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把`CustomerID`变量的数据类型转换为字符串

In [59]:
cleaned_date['CustomerID'] = cleaned_date['CustomerID'].astype(str)
cleaned_date['CustomerID']

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把`CustomerID`变量结尾的'.0'删除

In [61]:
cleaned_date['CustomerID'] = cleaned_date['CustomerID'].str.slice(0,-2) 
cleaned_date['CustomerID']

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把`Description`变量缺失的观察值删除

In [72]:
# cleaned_date = cleaned_date.dropna(subset=['Description'])
cleaned_date.dropna(subset=['Description'],inplace=True)


0

In [ ]:
 # 检查空值
cleaned_date['Description'].isnull().sum()

把`Country`变量值`"USA"`替换为`"United States"`  

In [79]:
cleaned_date[USA].replace('USA','United States',inplace=True)

In [88]:
# 检查是否全部修改成功
len(cleaned_date[cleaned_date['Country'] == 'USA'])

0

把`Country`变量值`"UK"`、`"U.K."`替换为`"United Kingdom"`

In [91]:
cleaned_date['Country'].replace(['UK',"U.K."],"United Kingdom",inplace=True)

In [93]:
# 检查是否全部修改成功
print(len(cleaned_date[cleaned_date['Country'] == 'UK']))
print(len(cleaned_date[cleaned_date['Country'] == 'U.K.']))

0
0


把`Quantity`变量值为负数的观察值删除  

In [95]:
cleaned_date = cleaned_date[cleaned_date['Quantity'] >= 0]

In [97]:
# 检查是否还有负值
len(cleaned_date[cleaned_date['Quantity']<0])

0

把`UnitPrice`变量值为负数的观察值删除

In [98]:
cleaned_date = cleaned_date[cleaned_date['UnitPrice'] >= 0]

In [99]:
# 检查是否还有负值
len(cleaned_date[cleaned_date['UnitPrice']<0])

0

## 保存清理后的数据

In [106]:
cleaned_date.to_csv('e_commerce_clearned.csv',index=False)

In [108]:
pd.read_csv('E:/file/data_analysis/python/小白玩转python/随课数据/数据评估和数据清洗篇/e_commerce_clearned.csv').head(10)

C:\Users\siweichao\AppData\Local\Temp\ipykernel_21928\1953442556.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv('E:/file/data_analysis/python/小白玩转python/随课数据/数据评估和数据清洗篇/e_commerce_clearned.csv').head(10)


InvoiceNo StockCode                          Description  ...  UnitPrice CustomerID         Country
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER  ...       2.55      17850  United Kingdom
1    536365     71053                  WHITE METAL LANTERN  ...       3.39      17850  United Kingdom
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER  ...       2.75      17850  United Kingdom
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE  ...       3.39      17850  United Kingdom
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.  ...       3.39      17850  United Kingdom
5    536365     22752         SET 7 BABUSHKA NESTING BOXES  ...       7.65      17850  United Kingdom
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER  ...       4.25      17850  United Kingdom
7    536366     22633               HAND WARMER UNION JACK  ...       1.85      17850  United Kingdom
8    536366     22632            HAND WARMER RED POLKA DOT  ...       1.85      17850  United Kingdom
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT  ...       1.69      13047  United Kingdom

[10 rows x 8 columns]